src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxVisitor.cs

```diff
@@ -16,11 +17,23 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </typeparam>
     public abstract partial class CSharpSyntaxVisitor<TResult>
     {
+        private const int MaxUncheckedRecursionDepth = Syntax.InternalSyntax.LanguageParser.MaxUncheckedRecursionDepth;
+        private int _recursionDepth;
+
         public virtual TResult Visit(SyntaxNode node)
         {
             if (node != null)
             {
-                return ((CSharpSyntaxNode)node).Accept(this);
+                _recursionDepth++;
+                if (_recursionDepth > MaxUncheckedRecursionDepth)
+                {
+                    PortableShim.RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+
+                var result = ((CSharpSyntaxNode)node).Accept(this);
+
+                _recursionDepth--;
```

Comment:
Should this be in a `try / finally` block?  The Parser doesn't use on because the code is wholly owned by us so if we throw then shame on us.  In this case though we're calling out to third party code so we can't guarantee they won't throw.